forgerock-auth/src/main/java/org/forgerock/android/auth/SecureCookieJar.kt

```diff
@@ -0,0 +1,160 @@
+        for (c in singleSignOnManager.cookies) {
+            val cookie = cookieMarshaller.unmarshal(c)
+            //Remove the same stored cookies
+            if (cookie != null && !contains(cookie, cookies)) {
+                cookiesContainer.add(cookie)
+            }
+        }
+        for (cookie in cookies) {
+            if (!isExpired(cookie)) {
+                cookiesContainer.add(cookie)
+            }
+        }
+        cache(cookiesContainer)
+        val updatedCookies: MutableSet<String> = HashSet()
+        for (cookie in cookiesContainer) {
+            updatedCookies.add(cookieMarshaller.marshal(cookie))
+        }
+        persist(updatedCookies)
+    }
+
+    private fun contains(input: Cookie, cookies: Collection<Cookie>): Boolean {
+        for (cookie in cookies) {
+            if (cookie.name == input.name && cookie.domain == input.domain && cookie.path == input.path) {
+                return true
+            }
+        }
+        return false
+    }
+
+    private fun filter(url: HttpUrl, cookies: Set<Cookie>): List<Cookie> {
+        val result = mutableListOf<Cookie>()
+        cookies.filter { !isExpired(it) }
+            .filter { it.matches(url) }
+            .toCollection(result)
+
+        return intercept(result.toImmutableList())
+    }
+
+    private fun isExpired(cookie: Cookie): Boolean {
+        return cookie.expiresAt < System.currentTimeMillis()
+    }
+
+    private fun cache(cookies: MutableSet<Cookie>) {
+        if (cacheIntervalMillis > 0) {
+            cacheRef.set(cookies)
+            worker.schedule({ cacheRef.set(null) }, cacheIntervalMillis, TimeUnit.MILLISECONDS)
+        }
+    }
+
+    private fun persist(cookies: Collection<String>) {
+        singleSignOnManager.persist(cookies)
+        FRLifecycle.dispatchCookiesUpdated(cookies)
+    }
+
+    class SecureCookieJarBuilder internal constructor() {
+        private var context: Context? = null
+        private var singleSignOnManager: SingleSignOnManager? = null
+        private var cacheIntervalMillis: Long? = null
+        fun context(context: Context?): SecureCookieJarBuilder {
+            this.context = context
+            return this
+        }
+
+        fun singleSignOnManager(singleSignOnManager: SingleSignOnManager?): SecureCookieJarBuilder {
+            this.singleSignOnManager = singleSignOnManager
+            return this
+        }
+
+        fun cacheIntervalMillis(cacheIntervalMillis: Long?): SecureCookieJarBuilder {
+            this.cacheIntervalMillis = cacheIntervalMillis
+            return this
+        }
+
+        fun build(): SecureCookieJar {
+            return SecureCookieJar(context, singleSignOnManager, cacheIntervalMillis)
+        }
+
+        override fun toString(): String {
+            return "SecureCookieJar.SecureCookieJarBuilder(context=" + context + ", singleSignOnManager=" + singleSignOnManager + ", cacheIntervalMillis=" + cacheIntervalMillis + ")"
```

Comment:
may be more cleaner  

 return "SecureCookieJar.SecureCookieJarBuilder(context=$context, singleSignOnManager=$singleSignOnManager, cacheIntervalMillis=$cacheIntervalMillis)"

Reply:
Updated